File: Canvas/TechDebt/Analytics.m
//
// Copyright (C) 2016-present Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, version 3 of the License.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//
    
    

#import "Analytics.h"
#import <Google/Analytics.h>

@implementation Analytics

+ (void)prepare {
    
    // Some open source clients might not use this at all
    NSURL *plistURL = [[NSBundle mainBundle] URLForResource:@"GoogleServices-Info" withExtension:@"plist"];
    NSDictionary *data = [NSDictionary dictionaryWithContentsOfURL:plistURL];
    if (!data) {
        return;
    }
    
    // Configure tracker from GoogleService-Info.plist. (included in Canvas Target)
    NSError *configureError;
    [[GGLContext sharedInstance] configureWithError:&configureError];
    NSAssert(!configureError, @"Error configuring Google services: %@", configureError);
}

+ (void)logScreenView:(NSString*)screenName {
    id<GAITracker> tracker = [[GAI sharedInstance] defaultTracker];
    [tracker set:kGAIScreenName value:screenName];
    [tracker send:[[GAIDictionaryBuilder createScreenView] build]];
}
@end

File: Core/Core/Analytics/Analytics.swift
//
// This file is part of Canvas.
// Copyright (C) 2019-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

import Foundation

public protocol AnalyticsHandler: AnyObject {
    func handleEvent(_ name: String, parameters: [String: Any]?)
}

public class Analytics: NSObject {
    public static var shared: Analytics = Analytics()

    public weak var handler: AnalyticsHandler?

    /**
     This method is mainly used to track user and application actions for usage statistics.
     */
    @objc
    public func logEvent(_ name: String, parameters: [String: Any]? = nil) {
        handler?.handleEvent(name, parameters: parameters)
    }

    public func logSession(_ session: LoginSession) {
        var defaults = SessionDefaults(sessionID: session.uniqueID)
        let tokenExpires = session.expiresAt != nil
        if defaults.tokenExpires == nil || defaults.tokenExpires != tokenExpires {
            let event = tokenExpires ? "auth_expiring_token" : "auth_forever_token"
            logEvent(event)
            defaults.tokenExpires = tokenExpires
        }
    }

    public static var analyticsBaseUrl: String {
        guard let session = AppEnvironment.shared.currentSession else {
            return ""
        }
        return session.baseURL.absoluteString
    }

    public static var analyticsBaseUrl: String {
        guard let session = AppEnvironment.shared.currentSession else {
            return ""
        }
        return session.baseURL.absoluteString
    }
}
